cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java

```diff
@@ -158,11 +169,21 @@ public void getOperationRun(
   @Path("/{id}/stop")
   public void failOperation(FullHttpRequest request, HttpResponder responder,
       @PathParam("namespace-id") String namespaceId,
-      @PathParam("id") String runId) {
-    // // TODO(samik, CDAP-20814) send the message to stop the operation
-    responder.sendString(HttpResponseStatus.OK,
-        String.format("Updated status for operation run %s in namespace '%s'.", runId,
-            namespaceId));
+      @PathParam("id") String runId)
+      throws OperationRunNotFoundException, IOException, BadRequestException, ForbiddenException {
+    checkSourceControlMultiAppFeatureFlag();
+    validateNamespaceId(namespaceId);
+    if (runId == null || runId.isEmpty()) {
+      throw new BadRequestException("runId cannot be empty");
+    }
+    // validate the run exists
+    OperationRunDetail runDetail =
+        operationLifecycleManager.getOperationRun(new OperationRunId(namespaceId, runId));
+    if (!runDetail.getRun().getStatus().canTransitionTo(OperationRunStatus.STOPPING)) {
+      throw new BadRequestException(String.format("Operation run in %s state cannot to be stopped",
+          runDetail.getRun().getStatus()));
+    }
+    statePublisher.publishStopping(new OperationRunId(namespaceId, runId));
```

Comment:
Just curious, where do we pick this message and perform the actual stopping of the running task?